packages/run-protocol/src/vpool-xyk-amm/stableSwapCurve.js

```diff
@@ -0,0 +1,388 @@
+ */
+const commonStablePrice = (
+  inputAmount,
+  poolAmounts,
+  tokenIndexFrom,
+  tokenIndexTo,
+  feeBasisPoints = 30n,
+) => {
+  let inputReserve = poolAmounts[tokenIndexFrom];
+  let outputReserve = poolAmounts[tokenIndexTo];
+  assert(
+    inputAmount.value > 0n,
+    X`inputValue ${inputAmount.value} must be positive`,
+  );
+  assert(
+    inputReserve.value > 0n,
+    X`inputReserve ${inputReserve.value} must be positive`,
+  );
+  assert(
+    outputReserve.value > 0n,
+    X`outputReserve ${outputReserve.value} must be positive`,
+  );
+  // Fee ratio calculation
+  const feeCutRatio = makeRatio(
+    BASIS_POINTS - feeBasisPoints,
+    inputAmount.brand,
+    BASIS_POINTS,
+    inputAmount.brand,
+  );
+  // Fee ratio multiplied by inputAmount to get inputAmount After fee cut
+  let inputAmountAfterFeeCut = floorMultiplyBy(inputAmount, feeCutRatio);
+  // Normalizing input amount according to pool value
+  inputAmountAfterFeeCut = AmountMath.make(
+    inputAmountAfterFeeCut.brand,
+    inputAmountAfterFeeCut.value * BASIS_POINTS,
+  );
+  const basisRatio = makeRatio(BASIS_POINTS * 100n, inputAmount.brand);
+  let inputAmountWithoutFeeCut = floorMultiplyBy(inputAmount, basisRatio);
+  // Normalizing the poolValue according to Basis_Points
+  let poolAmountsInBasisPoints = poolAmounts.map(amount => {
+    return floorMultiplyBy(
+      amount,
+      makeRatio(BASIS_POINTS * 100n, amount.brand),
+    );
+  });
+  const poolValues = poolAmountsInBasisPoints.map(amount => amount.value);
+  return { inputAmountAfterFeeCut, inputAmountWithoutFeeCut, poolValues };
+};
+/**
+ * Recompute the pool values after a swap is performed
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ *                                passed from the contract.
+ * @param {number} tokenIndexFrom - index of amount of inputReserve
+ *                                  in the reserves array.
+ * @param {number} tokenIndexTo - index of amount of outputReserve
+ *                                in the reserves array.
+ * @param {bigint} inputTokenValue - the value of swap in token passed in.
+ * @param {bigint} outputTokenValue - the value of swap out token returned
+ *                                    in exchange
+ * @returns {bigint[]} output - recomputed pool values
+ */
+const updatePoolValues = (
+  poolValues,
+  tokenIndexFrom,
+  inputTokenValue,
+  tokenIndexTo,
+  outputTokenValue,
+) => {
+  poolValues[tokenIndexFrom] -= inputTokenValue;
+  poolValues[tokenIndexTo] += outputTokenValue;
+  return poolValues;
+};
+
+/**
+ * Computes the Stable Swap invariant (D).
+ * @param {bigint[]} poolValues - Array of amounts of each asset.Which is
+ * passed from the contract.
+ * @returns {bigint} d - the current price, in value form
+ *
+ */
+export const getD = poolValues => {
+  const N_COINS = poolValues.length;
```

Comment:
We prefer ALL_CAPS for values [that don't vary from run to run](https://github.com/airbnb/javascript/#naming--uppercase). This variable's value is set each time `getD` is called.
```suggestion
  const nCoins = poolValues.length;
```

Oh,  and these should be bigints so we can work with ratios so we don' t lose precision when we do division.
```suggestion
  const nCoins = BigInt(poolValues.length);
```